projects/ngx-mapbox-gl/src/lib/source/raster-dem-source.component.ts

```diff
@@ -0,0 +1,88 @@
+import { ChangeDetectionStrategy, Component, Input, OnChanges, OnDestroy, OnInit, SimpleChanges } from '@angular/core';
+import { RasterDemSource } from 'mapbox-gl';
+import { fromEvent, Subscription } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { MapService } from '../map/map.service';
+
+@Component({
+  selector: 'mgl-raster-dem-source',
+  template: '',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class RasterDemSourceComponent implements OnInit, OnDestroy, OnChanges, RasterDemSource {
+  /* Init inputs */
+  @Input() id: string;
+
+  /* Dynamic inputs */
+  @Input() url?: RasterDemSource['url'];
+  @Input() tiles?: RasterDemSource['tiles'];
+  @Input() bounds?: RasterDemSource['bounds'];
+  @Input() minzoom?: RasterDemSource['minzoom'];
+  @Input() maxzoom?: RasterDemSource['maxzoom'];
+  @Input() tileSize?: RasterDemSource['tileSize'];
+  @Input() attribution?: RasterDemSource['attribution'];
+  @Input() encoding?: RasterDemSource['encoding'];
+
+  type: RasterDemSource['type'] = 'raster-dem'; // Just to make ts happy
+
+  private sourceAdded = false;
+  private sub = new Subscription();
+
+  constructor(private MapService: MapService) {}
+
+  ngOnInit() {
+    const sub1 = this.MapService.mapLoaded$.subscribe(() => {
+      this.init();
+      const sub = fromEvent(<any>this.MapService.mapInstance, 'styledata')
+        .pipe(filter(() => !this.MapService.mapInstance.getSource(this.id)))
+        .subscribe(() => {
```

Comment:
Avoid subscribe inside subscribe. Please use a rxjs operator to handle the flattening strategy.

Reply:
(or you can also separate this new component in another PR)

Reply:
I copied this logic from all the other sources actually. It was added in https://github.com/Wykks/ngx-mapbox-gl/commit/4c496299b62fd01a0387fdc61a42c9751b7157f0. The challenge with using `merge` or `concat` to flatten is that `this.MapService.mapInstance` is not set until `mapLoaded$` fires. We therefore have to wait before subscribing.

Reply:
Oh right. It's been two years since I actually wrote that... Fine then :+1: